hydro_cli/src/core/gcp.rs

```diff
@@ -0,0 +1,444 @@
+use std::{
+    net::SocketAddr,
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+use anyhow::Result;
+use async_channel::{Receiver, Sender};
+
+use async_ssh2_lite::{AsyncChannel, AsyncSession, SessionConfiguration};
+use async_trait::async_trait;
+use futures::{AsyncWriteExt, Future, StreamExt};
+use hydroflow::util::connection::BindType;
+use serde_json::json;
+use tokio::{net::TcpStream, sync::RwLock};
+
+use super::{
+    localhost::create_broadcast,
+    terraform::{TerraformOutput, TerraformProvider},
+    ConnectionType, Host, LaunchedBinary, LaunchedHost, ResourceBatch, ResourceResult,
+};
+
+struct LaunchedComputeEngineBinary {
+    _resource_result: Arc<ResourceResult>,
+    channel: AsyncChannel<TcpStream>,
+    stdin_channel: Sender<String>,
+    stdout_receivers: Arc<RwLock<Vec<Sender<String>>>>,
+    stderr_receivers: Arc<RwLock<Vec<Sender<String>>>>,
+}
+
+#[async_trait]
+impl LaunchedBinary for LaunchedComputeEngineBinary {
+    async fn stdin(&self) -> Sender<String> {
+        self.stdin_channel.clone()
+    }
+
+    async fn stdout(&self) -> Receiver<String> {
+        let mut receivers = self.stdout_receivers.write().await;
+        let (sender, receiver) = async_channel::unbounded::<String>();
+        receivers.push(sender);
+        receiver
+    }
+
+    async fn stderr(&self) -> Receiver<String> {
+        let mut receivers = self.stderr_receivers.write().await;
+        let (sender, receiver) = async_channel::unbounded::<String>();
+        receivers.push(sender);
+        receiver
+    }
+
+    async fn exit_code(&self) -> Option<i32> {
+        if self.channel.eof() {
+            self.channel.exit_status().ok()
+        } else {
+            None
+        }
+    }
+}
+
+struct LaunchedComputeEngine {
+    resource_result: Arc<ResourceResult>,
+    external_ip: String,
+    binary_counter: RwLock<usize>,
+}
+
+async fn async_retry<T, F: Future<Output = Result<T>>>(
+    thunk: impl Fn() -> F,
+    count: usize,
+) -> Result<T> {
+    for _ in 1..count {
+        let result = thunk().await;
+        if result.is_ok() {
+            return result;
+        } else {
+            tokio::time::sleep(std::time::Duration::from_millis(1000)).await;
+        }
+    }
+
+    thunk().await
```

Comment:
Any reason for not just doing this as sequential async code?

Reply:
Not sure I understand your question. The reason for the extra call at the end is so we can propagate errors back up to the user.